networkit/cpp/centrality/ComplexPaths.cpp

```diff
@@ -0,0 +1,203 @@
+    if (mode == Mode::singleNode && start == none)
+        ERROR("complexPathAlgorithm: provide node != none in Mode::singleNode.");
+    if (mode == Mode::allNodes && start != none)
+        WARN("complexPathAlgorithm: start node ignored in Mode::allNodes.");
+}
+
+void ComplexPathAlgorithm::normalize() {
+    if (mode != Mode::allNodes) {
+        WARN("complexPathAlgorithm: no normalization in Mode::singleNode.");
+        return;
+    }
+    normPaths = true;
+}
+
+void ComplexPathAlgorithm::run() {
+    if (mode == Mode::singleNode) {
+        complexPathGraph = complexPathsGraph(start, threshold, &adopters);
+    } else {
+        complexPathsLengths = complexPathLength(threshold);
+    }
+    hasRun = true;
+}
+
+vector<double> ComplexPathAlgorithm::getPLci() {
+    if (mode != Mode::allNodes)
+        WARN("complexPathAlgorithm[getPLci]: no results in Mode::singleNode.");
+    assureFinished();
+    return complexPathsLengths;
+}
+
+Graph ComplexPathAlgorithm::getComplexGraph() {
+    if (mode != Mode::singleNode)
+        WARN("complexPathAlgorithm[getComplexGraph]: no results in Mode::allNodes.");
+    assureFinished();
+    return complexPathGraph;
+}
+
+vector<node> ComplexPathAlgorithm::getAdopters() {
+    if (mode != Mode::singleNode)
+        WARN("complexPathAlgorithm[getAdopters]: no results in Mode::allNodes.");
+    assureFinished();
+    return adopters;
+}
+
+static void addNewEdge(Graph &g, node u, node v) {
+    if (g.hasEdge(u, v))
+        return;
+    g.addEdge(u, v);
+}
+
+Graph ComplexPathAlgorithm::complexPathsGraph(node seed, count threshold, vector<node> *adopters) {
+    const auto &g = inputGraph;
+    auto n = g.numberOfNodes();
+    Graph complex_g = Graph(n); // all nodes, no edges so far
+    vector<bool> activated(n);
+    vector<count> influence(n);
+
+    // activate threshold-1 random neighbors of seed:
+    auto needSeeds = threshold - 1;
+
+    vector<node> seeds;
+    seeds.reserve(4);
+
+    auto neighbors = Graph::NeighborRange<>(g, seed);
+
+    auto randomGen = Aux::Random::getURNG();
+    sample(neighbors.begin(), neighbors.end(), std::back_inserter(seeds), needSeeds, randomGen);
+    auto needMore = needSeeds - seeds.size();
+    if (needMore > 0) {
+        WARN("complexPathsGraph: ", seed, ": not enough direct neighbors.");
+        vector<node> moreSeeds;
+        for (auto u : neighbors) {
+            auto indirectNeighbors = Graph::NeighborRange<>(g, u);
+            for (auto in : indirectNeighbors) {
+                if (in != seed
+                    && std::find(moreSeeds.begin(), moreSeeds.end(), in) == moreSeeds.end())
+                    moreSeeds.push_back(in);
+            }
+            sample(moreSeeds.begin(), moreSeeds.end(), std::back_inserter(seeds), needMore,
+                   randomGen);
+            if (seeds.size() < needSeeds)
+                WARN("complexPathsGraph: ", seed, ": not enough indirect neighbors.");
```

Comment:
This is inefficient and overly complicated. We basically need to sample a fixed number of seeds from the nodes up to distance 2 from the seed. Why don't we:
- Sample from the neighbors of `seed` only if they are enough, otherwise
- Run a BFS that stops at level 2 and use two queues to keep all the nodes at distance 1 and 2 separate. Then insert all the nodes at level 1 and sample the remaining number of seeds from the nodes at level 2.

In this way the code would be simpler and we won't need to run all these `std::find` (which are linear searches).